pkg/gameserverallocations/controller.go

```diff
@@ -0,0 +1,394 @@
+	}
+
+	if ok, causes := oldGSA.ValidateUpdate(newGSA); !ok {
+		review.Response.Allowed = false
+		details := metav1.StatusDetails{
+			Name:   review.Request.Name,
+			Group:  review.Request.Kind.Group,
+			Kind:   review.Request.Kind.Kind,
+			Causes: causes,
+		}
+		review.Response.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "GameServerAllocation update is invalid",
+			Reason:  metav1.StatusReasonInvalid,
+			Details: &details,
+		}
+	}
+
+	return review, nil
+}
+
+// allocate allocated a GameServer from a given Fleet
+func (c *Controller) allocate(gsa *v1alpha1.GameServerAllocation) (*v1alpha1.GameServer, error) {
+	var allocation *v1alpha1.GameServer
+	// can only allocate one at a time, as we don't want two separate processes
+	// trying to allocate the same GameServer to different clients
+	c.allocationMutex.Lock()
+	defer c.allocationMutex.Unlock()
+
+	// make sure we have the most up to date view of the world
+	if !cache.WaitForCacheSync(c.stop, c.gameServerSynced) {
+		return allocation, errors.New("error syncing GameServer cache")
+	}
+
+	var comparator findComparator
+
+	switch gsa.Spec.Scheduling {
+	case v1alpha1.Packed:
+		comparator = packedComparator
+	case v1alpha1.Distributed:
+		comparator = distributedComparator
+	}
+
+	allocation, err := c.findReadyGameServerForAllocation(gsa, comparator)
+	if err != nil {
+		return allocation, err
+	}
+
+	gsCopy := allocation.DeepCopy()
+	gsCopy.Status.State = v1alpha1.GameServerStateAllocated
+
+	c.patchMetadata(gsCopy, gsa.Spec.MetaPatch)
+
+	patch, err := allocation.Patch(gsCopy)
+	if err != nil {
+		return allocation, err
+	}
+
+	gs, err := c.gameServerGetter.GameServers(gsCopy.ObjectMeta.Namespace).
+		Patch(gsCopy.ObjectMeta.Name, types.JSONPatchType, patch)
+	if err != nil {
+		return gs, errors.Wrapf(err, "error updating GameServer %s", gsCopy.ObjectMeta.Name)
+	}
+	c.recorder.Event(gs, corev1.EventTypeNormal, string(gs.Status.State), "Allocated")
+
+	return gs, nil
+}
+
+// patch the labels and annotations of an allocated GameServer with metadata from a GameServerAllocation
+func (c *Controller) patchMetadata(gs *v1alpha1.GameServer, fam v1alpha1.MetaPatch) {
+	// patch ObjectMeta labels
+	if fam.Labels != nil {
+		if gs.ObjectMeta.Labels == nil {
+			gs.ObjectMeta.Labels = make(map[string]string, len(fam.Labels))
+		}
+		for key, value := range fam.Labels {
+			gs.ObjectMeta.Labels[key] = value
+		}
+	}
+	// apply annotations patch
+	if fam.Annotations != nil {
+		if gs.ObjectMeta.Annotations == nil {
+			gs.ObjectMeta.Annotations = make(map[string]string, len(fam.Annotations))
+		}
+		for key, value := range fam.Annotations {
+			gs.ObjectMeta.Annotations[key] = value
+		}
+	}
+}
+
+// syncDelete takes unallocated GameServerAllocations, and deletes them!
+func (c *Controller) syncDelete(key string) error {
+	c.logger.WithField("key", key).Info("Deleting gameserverallocation")
```

Comment:
Does this means the `GameServerAllocation` is added with a status `GameServerAllocationUnAllocated` but deleted right after. Asking question mostly for metrics implementation.

Reply:
Yeah - I couldn't think of a better way to handle UnAllocated GameServerAllocations - since there is no GameServer that will eventually get deleted, it need to be returned in the webhook, but eventually deleted, otherwise it would hang around forever.